### PR TITLE
Starting the document on TLA+ idioms

### PR DESCRIPTION
**Pull request by konnov**
This document has been living in a separate branch for several months. Let's start writing it. The first two idioms are in.

### CODE REVIEW

File: docs/idiomatic/README.md [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+1. understood by distributed system engineers,
+1. understood by verification engineers, and
+1. understood by automatic analysis tools such as the Apalache model checker.
+
+If you believe, that the above points are contradictory when put together, it is
+to some extent true. TLA+ is an extremely general specification language. As a
+result, it is easy to write a short specification that leaves puzzled a human
+reader. It is even easier to write a (syntactically correct) specification that
+turns to dust any program reasoning about TLA+.
```
Comment (shonfeder) [inline docs/idiomatic/README.md:21]:
```suggestion
turns any program trying to reason about TLA+ to dust. 
```

File: docs/idiomatic/README.md [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+specifications that are:
+
+1. understood by distributed system engineers,
+1. understood by verification engineers, and
+1. understood by automatic analysis tools such as the Apalache model checker.
+
+If you believe, that the above points are contradictory when put together, it is
+to some extent true. TLA+ is an extremely general specification language. As a
+result, it is easy to write a short specification that leaves puzzled a human
+reader. It is even easier to write a (syntactically correct) specification that
```
Comment (shonfeder) [inline docs/idiomatic/README.md:20]:
```suggestion
result, it is easy to write a short specification that leaves a human reader
puzzled . It is even easier to write a (syntactically correct) specification that
```

File: docs/idiomatic/README.md [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+turns to dust any program reasoning about TLA+. 
+
+Nevertheless, we find TLA+ quite useful when writing concise specifications of
+distributed protocols at [Informal Systems]. Other specification languages --
+especially, those designed for software verification -- would require us to
+introduce unnecessary book-keeping details that would both obfuscate the
+protocols and make their verification harder. However, we do not always need
+_"all power of mathematics"_, so we find it useful to introduce additional
```
Comment (shonfeder) [inline docs/idiomatic/README.md:28]:
```suggestion
_"all the power of mathematics"_, so we find it useful to introduce additional
```

File: docs/idiomatic/README.md [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+structure in TLA+ specifications.
+
+Below, we summarize the idioms that help us in maintaining that structure.  As
+a bonus, these idioms usually aid the Apalache model checker in analyzing the
+specifications. Our idioms are quite likely different from the original ideas
+of [Leslie Lamport] (the author of TLA+).
+So it is useful to read Lamport's [Specifying Systems]. Importantly, these are
+_idioms_, not the rules set in stone. If you believe that one of those idioms
```
Comment (shonfeder) [inline docs/idiomatic/README.md:36]:
```suggestion
_idioms_, not rules set in stone. If you believe that one of those idioms
```

File: docs/idiomatic/README.md [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+does not work for you in your specific setting, don't follow it.
+
+If this is the first page where you encounter the word "TLA+", we do not
+recommend continue reading the material. It is better to start with [The TLA+
```
Comment (shonfeder) [inline docs/idiomatic/README.md:40]:
```suggestion
recommend that you continue to read the material. It is better to start with [The TLA+
```

File: docs/idiomatic/README.md [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+Video Course](http://lamport.azurewebsites.net/video/videos.html) by Leslie
+Lamport.  Once you have understood the basics and tried the language, it makes
+sense to ask the question: "How do I write a specification that other people
+understand?".  We believe that many TLA+ users reinvent the rules that are
+similar to our idioms.  By providing you with a bit of thought, we hope to
```
Comment (shonfeder) [inline docs/idiomatic/README.md:45]:
```suggestion
understand?".  We believe that many TLA+ users reinvent rules that are
similar to our idioms.  By providing you with a bit of guidance, we hope to
```

File: docs/idiomatic/README.md [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+reduce your discomfort when learning TLA+.
```
Comment (shonfeder) [inline docs/idiomatic/README.md:46]:
```suggestion
reduce your discomfort when learning more advanced TLA+.
```

**Review by shonfeder — APPROVED** [review]

File: docs/idiomatic/000keep-minimum-state-variables.md [hunk 8]
```diff
@@ -0,0 +1,150 @@
+# Idiom 0: Keep state variables to the minimum
```
[the remark this answers is not in the document]
Comment (konnov) [reply docs/idiomatic/000keep-minimum-state-variables.md:1]: Yes. I think this is also my intuition.

File: docs/idiomatic/000keep-minimum-state-variables.md [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+if we want to closely monitor the values of the computed expressions, when using
+the specification for model-based testing.
+
+## Example
+
+Consider the following implementation of [Bubble sort] in Python:
+
+```python
+    my_list = [5, 4, 3, 8, 1]
+    changed = True
```
Comment (konnov) [inline docs/idiomatic/000keep-minimum-state-variables.md:47]:
```suggestion
    finished = False
```

File: docs/idiomatic/000keep-minimum-state-variables.md [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+    my_list_len = len(my_list)  # cache the length
+    while changed:
+        changed = False
```
Comment (konnov) [inline docs/idiomatic/000keep-minimum-state-variables.md:50]:
```suggestion
    while not finished:
        finished = True
```

File: docs/idiomatic/000keep-minimum-state-variables.md [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+        for i in range(1, my_list_len):
+            current = my_list[i]
+            if prev <= current:
+                # save current for the next iteration
+                prev = current
+            else:
+                # swap the elements
+                my_list[i - 1] = current
+                my_list[i] = prev
+                changed = True
```
Comment (konnov) [inline docs/idiomatic/000keep-minimum-state-variables.md:62]:
```suggestion
                finished = False
```

File: docs/idiomatic/000keep-minimum-state-variables.md [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+  - `prev` to cache the previously accessed element of the list,
+    in order to minimize the number of list accesses,
+  - `current` to cache the iterated element of the list.
+
+In TLA+, one usually does not introduce local variables for the intermediate
+results of the computation, but rather introduces variables to represent the
+essential part of the algorithm state. (While we have spent some time on code
+optimization, we might have missed the fact that our sorting algorithm is not
+as good as [Quicksort].) In the above example, the essential variables are
+`finished` and `my_list`.
```
[the remark this answers is not in the document]
Comment (konnov) [reply docs/idiomatic/000keep-minimum-state-variables.md:77]: Good point. I guess, I have refactored the TLA+ spec without looking back at the python code. Updated the python code.

File: docs/idiomatic/001assignments.md [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+   to the primed variables.
+
+ - Non-determinism is clearly isolated in existential choice: `\E y \in S: x' := y`.
+   If there is no existential choice, the assignment is deterministic.
+
+ - When the existential form is used, the range of the values is clearly indicated.
+   This is in contrast to the negated form such as: `~(x' = 10)`.
+
+ - TLC treats the expressions of the form `x' = e` and `x' \in S` as assignments,
+   as long as `x'` is not bound to a value.
```
[the remark this answers is not in the document]
Comment (konnov) [reply docs/idiomatic/001assignments.md:118]: Neither it is clear to me now... I guess, what I wanted to say is that it is consistent with the approach of TLC. It also cannot treat arbitrary constraints over primes.

File: docs/idiomatic/001assignments.md [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+## Disadvantages
+
+  - Replacing `x' \in S` with `\E y \in S: x' := y` makes the specification a bit larger.
+
+## Example
+
+The following example [deliver.tla](./example/deliver.tla) demonstrates how
+one can clear mark assignments.
+
+```tla
```
[the remark this answers is not in the document]
Comment (konnov) [reply docs/idiomatic/001assignments.md:133]: Yes. This is a good idea! Maybe in a follow-up PR. This document will require many iterations.

File: docs/idiomatic/002primes.md [hunk 15]
```diff
@@ -0,0 +1,133 @@
+# Idiom 2: Apply primes only to state variables
+
+(_Until you learn how prime actually works!_)
+
+## Description
+
+In many formal languages, the notation `x'` denotes the value that a variable
```
[the remark this answers is not in the document]
Comment (konnov) [reply docs/idiomatic/002primes.md:7]: In model checking, this notation became the de-facto standard. People are using it without knowing the origins.

File: docs/idiomatic/README.md [hunk 16]
```diff
@@ -0,0 +1,85 @@
+# Idiomatic TLA+
+
+**Authors:** Igor Konnov + (who likes to contribute?)
```
Comment (konnov) [inline docs/idiomatic/README.md:3]:
```suggestion
**Authors:** Shon Feder, Igor Konnov + (who likes to contribute?)
```

File: docs/idiomatic/000keep-minimum-state-variables.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (konnov) [reply docs/idiomatic/000keep-minimum-state-variables.md:1]: I will think about where to add it. If you have a good phrase for it, just add it.

**Comment by lemmy** [comment]
Perhaps, the definition of `NotifyOther` below could serve as an example of when to add additional variables to make a spec human-friendly: 

```tla
--------------------------- MODULE BlockingQueueFair ---------------------------

...

VARIABLES buffer, waitSeq
vars == <<buffer, waitSeq>>


Range(f) == { f[i]: i \in DOMAIN f }
waitSet == Range(waitSeq)

RunningThreads == (Producers \cup Consumers) \ waitSet

NotifyOther(t) == 
          LET S == IF t \in Producers THEN waitSet \ Producers ELSE waitSet \ Consumers
              R == SelectSeq(waitSeq, LAMBDA u:u \in S)
              T == SelectSeq(waitSeq, LAMBDA u:u \notin S)
          IN IF R # <<>>
             THEN waitSeq' = Tail(R) \o T
             ELSE UNCHANGED waitSeq

(* @see java.lang.Object#wait *)
Wait(t) == /\ waitSeq' = Append(waitSeq, t)
           /\ UNCHANGED <<buffer>>
           
-----------------------------------------------------------------------------

Put(t, d) ==
   \/ /\ Len(buffer) < BufCapacity
      /\ buffer' = Append(buffer, d)
      /\ NotifyOther(t)
   \/ /\ Len(buffer) = BufCapacity
      /\ Wait(t)
      
Get(t) ==
   \/ /\ buffer # <<>>
      /\ buffer' = Tail(buffer)
      /\ NotifyOther(t)
   \/ /\ buffer = <<>>
      /\ Wait(t)

...
=============================================================================
```

In contrast, [this is the variant with one more variable](https://github.com/lemmy/BlockingQueue/blob/3a66f46f6f5703f2863f71baaf0aedaaee58836f/BlockingQueueSplit.tla#L16-L51) that separates `wait` into `Producers` and `Consumers`.

File: docs/idiomatic/001assignments.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (shonfeder) [reply docs/idiomatic/001assignments.md:123]: Perhaps this line should be moved to another section then?

**Comment by konnov** [comment]
Nice example, @lemmy. It indeed looks cleaner with one more variable. By the way, I am also violating this rule from time to time :-)

File: docs/idiomatic/000keep-minimum-state-variables.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (shonfeder) [reply docs/idiomatic/000keep-minimum-state-variables.md:1]: How about right here?

```suggestion
# Idiom 0: Keep state variables to the minimum

In imperative programming, it is common to use mutable variable assignments liberally, but to exercise caution whenever mutable variables have a global scope. In TLA+, mutable variables are always global, so it is important to use them carefully and in a way that accurately reflects the global state of the system you are specifying.  
```

File: docs/idiomatic/001assignments.md [hunk 17]
```diff
@@ -0,0 +1,190 @@
+# Idiom 1: Update state variables with assignments
+
+## Description
+
+The idiom "[Keep state variables to the
+minimum](000keep-minimum-state-variables.md)" tells us to store the absolutely
+necessary minimum in the state variables. By following this idiom, we develop
```
Comment (adizere) [inline docs/idiomatic/001assignments.md:7]: This is a bit of a mouthful.

```suggestion
minimum](000keep-minimum-state-variables.md)" tells us to store the minimum
necessary state variables. By following this idiom, we develop
```

File: docs/idiomatic/000keep-minimum-state-variables.md [hunk 18]
```diff
@@ -0,0 +1,149 @@
+# Idiom 0: Keep state variables to the minimum
+
+## Description
+
+_A good TLA+ specification minimizes the computation state and makes it visible_.
+
+TLA+ does not have special syntax for assignments to the variables.  For a good
```
Comment (adizere) [inline docs/idiomatic/000keep-minimum-state-variables.md:7]:
```suggestion

TLA+ does not have special syntax for variable assignment.  For a good
```

Comment (adizere) [reply docs/idiomatic/000keep-minimum-state-variables.md:7]: BTW, this statement is controversial. TLA+ does have variable assignment, which is what idiom 001 states explicitly:

> TLA+ uses the same glyph, `=`, for assignment, asserting equality, and binding variables.

Perhaps we can rephrase the whole sentence to be consistent with idiom 001.

File: docs/idiomatic/001assignments.md [hunk 19]
```diff
@@ -0,0 +1,190 @@
...
+the specification by writing constraints over the primed variables.
+
+TLA+ comes with a great freedom of expressing constraints over variables.
+While we love TLA+ for that freedom, we believe that constraints over primed
+variables are sometimes confusing.
+TLA+ uses the same glyph, `=`, for assignment, asserting equality, and binding variables. But these are very different operations and have different semantics.
```
Comment (adizere) [inline docs/idiomatic/001assignments.md:13]:
```suggestion
TLA+ uses the same glyph, `=` for three separate purposes: assignment, asserting equality, and binding variables. But these are very different operations and have different semantics.
```

**Review by adizere — COMMENTED** [review]
Very cool!

I initially wanted to ask if this repo is the right place to collect the idioms (VDD might have been better), but I understand that some idioms are actually Apalache-specific.

File: docs/idiomatic/000keep-minimum-state-variables.md (hunk 18 above)
Comment (shonfeder) [reply docs/idiomatic/000keep-minimum-state-variables.md:7]: Perhaps something like this?

```suggestion

TLA+ does not have special syntax that distinguishes mutable variable assignment from equality constraints.  For a good
```

File: docs/idiomatic/000keep-minimum-state-variables.md [hunk 20]
```diff
@@ -0,0 +1,151 @@
...
+By avoiding auxiliary state variables, we localize the updates to the state.
+This improves specification readability. It also helps the tools, as large parts
+of the specification become deterministic.
+
+## Disadvantages
+
+Sometimes, we have to expose the internals of the computation. For instance,
+if we want to closely monitor the values of the computed expressions, when using
+the specification for model-based testing.
+
```
Comment (konnov) [inline docs/idiomatic/000keep-minimum-state-variables.md:42]:
```suggestion

Sometimes, we have to break this idiom to make the specification more readable. Here is an example by Markus Kuppe. The specification of [BlockingQueue](https://github.com/lemmy/BlockingQueue/blob/3a66f46f6f5703f2863f71baaf0aedaaee58836f/BlockingQueueSplit.tla#L16-L51) that has one more variable is easier to read than [the original specification](https://github.com/lemmy/BlockingQueue/blob/3a66f46f6f5703f2863f71baaf0aedaaee58836f/BlockingQueue.tla) with a minimal number of variables.
```

File: docs/idiomatic/001assignments.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (konnov) [reply docs/idiomatic/001assignments.md:123]: Let's just keep it and update the document later.

**Comment by konnov** [comment]
I will merge this PR so it becomes more visible. Feel free to propose changes.

**Comment by lemmy** [comment]
Perhaps, auxiliary (history/prophecy) variables deserve mention too.

**Pull request merged**